doc/specification.md

```diff
@@ -35,62 +35,126 @@ TLS
 
 # 4 Service Definition
 
-The gRIBI service is defined as a single gRPC service, with three RPCs:
+The gRIBI service is a single gRPC service currently defined in [`gribi.proto`](https://github.com/openconfig/gribi/blob/master/v1/proto/service/gribi.proto). It includes three RPCs:
 
- * `Modify` - a bidirectional stream RPC which is used by the client to inject
-   routing entries to the server - each part of an individual operation. The 
-   server responds asynchronously to these operations with an acknowledgement
-   mode, based on the operating mode of the RPC.
+ * `Modify` - used by the clients to modify the device's RIB.
  * `Get` - a server streaming RPC which can be used by a client to retrieve the
    current set of installed gRIBI entries.
  * `Flush` - a unary RPC that is used as a low-complexity means to remove
    entries from a server.
 
-## 4.1 The `Modify` RPC
+IANA has reserved [TCP port 9340](https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml?search=9340#Google_Networking) for gRIBI service.
 
-High-level description of `Modify` semantics.
+## 4.1 `Modify`
 
-* `ADD`
-  * repeated `ADD` translates to `MODIFY`
-* `MODIFY`
-  * must fail if referenced object does not exist.
-* `DELETE`
-  * error handling for missing entries
-  * error handling for forward references
+The `Modify` provides a bidirectional streaming RPC for clients to modify the device's RIB. A client sends `ModifyRequest` messages that contains a set of `AFTOperations` to the device. The device process the received requests and responds them asynchronously.
+
+### 4.1.1 Client-Server Session Neogotiation
+
+A gRIBI client is identified by gRPC connection, i.e., if a connection drops and reconnect with the same `election_id` value, it will be considered as another client.
+
+Before a client starting sending `AFTOperations`, it should reach agreement with the device regarding the following session parameters
+* Redundancy Mode (defined in x.y.z)
+* Persistent Mode (defined in x.y.z)
+* Acknoledge Mode (defined in x.y.z)
+
+A client starts the neogotiation process by sending the first `ModifyRequest` message with only `params` populated. `params` MUST NOT be sent more than once during the lifetime of the RPC session. All clients MUST send the same values of all the attributes of `params`. If the device can process and support the requesed parameters, it should respond with `ModifyResponse` that has `session_params_result.status = OK`. Otherwise, the device should close the `Modify` RPC and set the generic gRPC [`Status.code`](https://github.com/googleapis/googleapis/blob/master/google/rpc/status.proto) per the following scenarios:
```

Comment:
```suggestion
A client starts the negotiation process by sending the first `ModifyRequest` message with only `params` populated. `params` MUST NOT be sent more than once during the lifetime of the RPC session. All clients MUST send the same values of all the attributes of `params`. If the device can process and support the requesed parameters, it should respond with `ModifyResponse` that has `session_params_result.status = OK`. Otherwise, the device should close the `Modify` RPC and set the generic gRPC [`Status.code`](https://github.com/googleapis/googleapis/blob/master/google/rpc/status.proto) per the following scenarios:
```